README.md

```diff
@@ -2,18 +2,126 @@
 
 A [Language Server Protocol](https://microsoft.github.io/language-server-protocol/) base for developing any LSP implementation in Clojure.
 
-## Server
+[![Clojars Project](https://img.shields.io/clojars/v/com.github.clojure-lsp/lsp4clj.svg)](https://clojars.org/com.github.clojure-lsp/lsp4clj)
 
-[![Clojars Project](https://img.shields.io/clojars/v/com.github.clojure-lsp/lsp4clj-server.svg)](https://clojars.org/com.github.clojure-lsp/lsp4clj-server)
+lsp4clj reads and writes from stdio, parsing JSON-RPC according to the LSP spec. It provides tools to allow server implementors to receive, process, and respond to any of the methods defined in the LSP spec, and to send their own requests and notifications to clients.
 
-The lsp4clj-server has the necessary integration with Input/Output, LSP-JSON parsing, allowing for users of this lib to just code the entrypoints of each LSP method.
+## Usage
 
-## Protocols
+### Create a server
 
-[![Clojars Project](https://img.shields.io/clojars/v/com.github.clojure-lsp/lsp4clj-protocols.svg)](https://clojars.org/com.github.clojure-lsp/lsp4clj-protocols)
+To initialize a server that will read from stdin and write to stdout:
 
-The lsp4clj-protocols contains only the protocols/interfaces for servers that want to extend the official LSP protocol to provide more features, also, it is used by lsp4clj-server itself.
+```clojure
+(lsp4clj.server/stdio-server {:in System/in, :out System/out})
+```
 
-## Known LSPs users
+The returned server will have a core.async `:log-ch`, from which you can read server logs (vectors beginning with a log level).
+
+```clojure
+(async/go-loop []
+  (when-let [[level & args] (async/<! (:log-ch server))]
+    (apply logger/warn level args)
+    (recur)))
+```
+
+### Receive messages
+
+To receive messages from a client, lsp4clj defines a pair of multimethods, `lsp4clj.server/receive-notification` and `lsp4clj.server/receive-request` that dispatch on the method name (as defined by the LSP spec) of an incoming JSON-RPC message.
+
+Server implementors should create `defmethod`s for the messages they want to process. (Other methods will be logged and responded to with a generic "Method not found" response.)
+
+These `defmethod`s receive 3 arguments, the method name, a "context", and the `params` of the [JSON-RPC request or notification object](https://www.jsonrpc.org/specification#request_object). The keys of the params will have been converted (recursively) to kebab-case keywords. Read on for an explanation of what a "context" is and how to set it.
```

Comment:
One thing I think we lost using multimethods is that we can't require that clients implement specific methods like we had on protocol, the LSP spec has some methods that all server should provide, they are required methods, like initialize, diagnostics, definition etc. I'm ok for now with this, we can think on how to make required some methods in the future

Reply:
To be pedantic, you're not _required_ to implement every method on a protocol, the docs just say that you _should_. If you don't, you don't get any compile time errors, though you do get a runtime error if you try to call the method you didn't define.

But I understand what you're saying—a protocol is a stronger indication of which methods you're expected to implement.

This is just one of the drawbacks to the multimethods. They also don't give you anywhere to hang extra data about handlers. I can imagine wanting to flag certain messages as cancellable, give them validators to apply to their input or output, or more generally give them middleware style wrappers, etc.

I think the best way to think about this is that one of the typical responsibilities of a server is to route requests. The multimethods work for that, but reitit, Compojure, et al. address the fact that there's more you might want to do with routing.

On the other hand, I haven't run into anything on the clojure-lsp side yet where I really regret the simplicity of the multimethods.

So, I guess my feeling is that multimethods might not be the best design long-term, but I'm not sure I have enough information to commit to another design either.

I've vaguely considered expressing the "router" as a hashmap, keyed by method name, which you pass in to the server when you create it. The server could validate that certain keys were present. That would also give us lots of flexibility about what the values of the hashmap are. They could be functions to call as the handler (much like a multimethod, which maps dispatch values to functions). Or they could be other hashmaps, with a handler function in each. We could add other data over time. If you feel strongly that this is a better solution, I can work on it.

Reply:
> To be pedantic, you're not required to implement every method on a protocol, the docs just say that you should. If you don't, you don't get any compile time errors, though you do get a runtime error if you try to call the method you didn't define.

yeah, the issue is that most clients expect that some methods are always available like diagnostics, but I agree it's best to be a little bit more loose than strict for now.

> Or they could be other hashmaps, with a handler function in each. We could add other data over time. If you feel strongly that this is a better solution, I can work on it.

I think it would work well, but I'm ok trying the multimethods for now

Reply:
I'm also split on this. On the one hand, multimethods bring great simplicity, but our tooling for them is a lot worse. Previously I could find the implementation of the specific message handler with a "find reference", that's gone with multimethods. Additionally, there's no indication of all possible methods I could implement.
I still like the multimethod approach, but maybe this will drive us to add better multimethod stuff into clojure-lsp 😄

Reply:
@Cyrik we improved a lot defmethods support on clojure-lsp, doesn't find implementations work for that case?

Reply:
I'd say there are two parts to discovery in this case... What has been implemented? And what should be implemented?

For me, in Emacs, `find-implementation` works well for finding what has been implemented.

<img width="687" alt="Screen Shot 2022-07-31 at 2 25 11 PM" src="https://user-images.githubusercontent.com/6774/182046142-1a751da9-1b0f-489f-ae1a-df676aa6778e.png">

In the last release we also improved the way defmethods appear in `document-symbols` (which editors use to make a sidebar with all the file's functions) so that's another way to figure out what's been implemented.

<img width="359" alt="Screen Shot 2022-07-31 at 2 25 48 PM" src="https://user-images.githubusercontent.com/6774/182046806-3a6c666a-d156-4404-8b2d-c7013ba20738.png">

But as @ericdallo mentioned originally, the multimethods leave something to be desired as far as answering what should be implemented. In my opinion, the protocols did a slightly better job of that, but only marginally. The protocols were essentially various lists of things that clojure-lsp had implemented, but required and optional features were mixed together. So, I'd argue that the true authority on what should be implemented was and still is the LSP spec.

But I think it'll be interesting to explore whether lsp4clj can improve that aspect of discoverability over time. Having the "routes" be a map of handlers would probably mean the editors would have even less tooling to help you find code. But lsp4clj could provide more feedback about whether your routes were valid.

Anyway, I'm planning to leave the multimethods as is, for now. IIRC pedestal went through 3 or 4 different router implementations before they finally found one that was a good mix of declarative, fast and flexible, and I think that's the one most projects use now. So, we're allowed to get it wrong several more times before we get it right!

Reply:
The thing I'm missing from defmethods is a way of seeing references to a specific method. I know that this isn't possible in the general case, which is the whole point of defmethods, but often enough there are hardcoded calls to a specific method that could be shown as a direct reference. Something like `(receive-request "shutdown")`.
The other thing is completeness, at least in this case.

But in I also fully agree with @mainej that we should leave this as is for now since these are very minor gripes with the system.

Reply:
@Cyrik let me see if I follow...

Suppose we have this multimethod:
```clojure
(defmulti caps identity)

(defmethod caps "a" [_] "A")
(defmethod caps "b" [_] "B")
```

At call sites like the following, you know for certain that `(defmethod caps "a" ...)` is being called, not `(defmethod caps "b" ...)`.
```clojure
(caps "a")
```
In theory `clojure-lsp` could provide navigation tools to link the `a` defmethod to its usage.

Of course this isn't possible when it's ambiguous which defmethod is being called:
```clojure
(let [lower (rand-nth ["a" "b"])]
  (caps lower))
```
In these situations `clojure-lsp` couldn't help.

Assuming I'm following, I agree, that makes it harder to understand how multimethods, or really any kind of polymorphism, is being used. Polymorphism limits tooling, and that's true of all editors and languages. Maybe the next generation of tooling will offer more insight into polymorphic calls!

For these particular `lsp4clj` multimethods—`receive-request` and `receive-notification`—I'd say the problem isn't very big. There's only one place that calls `receive-request` and one that calls `receive-notification` (both in `lsp4clj`), so there's not much need to find the call sites.

I think there's a stronger argument that the other main entrypoints into `lsp4clj`—`send-request` and `send-notification`—exhibit a similar problem, even though they aren't multimethods. They accept the JSON-RPC method as an argument, so it could be difficult to find all the places that, for example, call `(lsp.server/send-notification server "$/progress" message)`. In `clojure-lsp` we've worked around this. We have only one place that calls `$/progress`, a function called `publish-progress`. If we want to find all the call sites, we can use `find-references` on `publish-progress`. And in fact, all the calls to `send-request` and `send-notification` are in similar wrapper functions, so it's pretty easy to find references.

Reply:
@mainej Yeah that's the exact case I'm talking about. In my experience, most multimethods are built in this way (using the first param for dispatch). But all of this is very much unrelated so we should probably stop the topic 😄 

To your second point: you're correct that this happens more often, I just didn't notice the change yet. Previously it was easy to find references on the `ILSPProducer` methods and you'll now have to do the work yourself. Maybe this could be pulled into lsp4clj as optional entry points. But I know that @ericdallo and I have differing opinions on how much lsp4clj should provide, so I'm also happy to just do that on our end. At some point, I want to publish a library with all the extra utility stuff around lsp4clj anyway (file-management, URL handling, and so on).

Reply:
we could open a new discussion on lsp4clj to see if we could provide something that help you like a different project using lsp4clj under the hood, I think discussing the problem a little bit more would make us understand better the problem

Reply:
https://github.com/clojure-lsp/lsp4clj/issues/14